datafusion/src/optimizer/simplify_expressions.rs

```diff
@@ -1191,212 +1178,171 @@ mod tests {
         assert_eq!(
-            (lit(ScalarValue::Boolean(None)).not_eq(lit(ScalarValue::Boolean(None))))
-                .rewrite(&mut rewriter)?,
+            do_simplify(
+                lit(ScalarValue::Boolean(None)).not_eq(lit(ScalarValue::Boolean(None)))
+            ),
             lit(ScalarValue::Boolean(None)),
         );
 
         // x != null is always null
         assert_eq!(
-            (col("c2").not_eq(lit(ScalarValue::Boolean(None)))).rewrite(&mut rewriter)?,
+            do_simplify(col("c2").not_eq(lit(ScalarValue::Boolean(None)))),
             lit(ScalarValue::Boolean(None)),
         );
 
         // null = x is always null
         assert_eq!(
-            (lit(ScalarValue::Boolean(None)).eq(col("c2"))).rewrite(&mut rewriter)?,
+            do_simplify(lit(ScalarValue::Boolean(None)).eq(col("c2"))),
             lit(ScalarValue::Boolean(None)),
         );
-
-        Ok(())
     }
 
     #[test]
-    fn simplify_expr_eq() -> Result<()> {
+    fn simplify_expr_eq() {
         let schema = expr_test_schema();
-        let mut rewriter = Simplifier {
-            schemas: vec![&schema],
-        };
-
-        assert_eq!(col("c2").get_type(&schema)?, DataType::Boolean);
+        assert_eq!(col("c2").get_type(&schema).unwrap(), DataType::Boolean);
 
         // true = ture -> true
-        assert_eq!((lit(true).eq(lit(true))).rewrite(&mut rewriter)?, lit(true),);
+        assert_eq!(do_simplify(lit(true).eq(lit(true))), lit(true));
 
         // true = false -> false
-        assert_eq!(
-            (lit(true).eq(lit(false))).rewrite(&mut rewriter)?,
-            lit(false),
-        );
+        assert_eq!(do_simplify(lit(true).eq(lit(false))), lit(false),);
 
         // c2 = true -> c2
-        assert_eq!((col("c2").eq(lit(true))).rewrite(&mut rewriter)?, col("c2"),);
+        assert_eq!(do_simplify(col("c2").eq(lit(true))), col("c2"));
 
         // c2 = false => !c2
-        assert_eq!(
-            (col("c2").eq(lit(false))).rewrite(&mut rewriter)?,
-            col("c2").not(),
-        );
-
-        Ok(())
+        assert_eq!(do_simplify(col("c2").eq(lit(false))), col("c2").not(),);
     }
 
     #[test]
-    fn simplify_expr_eq_skip_nonboolean_type() -> Result<()> {
+    fn simplify_expr_eq_skip_nonboolean_type() {
         let schema = expr_test_schema();
-        let mut rewriter = Simplifier {
-            schemas: vec![&schema],
-        };
 
-        // When one of the operand is not of boolean type, folding the other boolean constant will
-        // change return type of expression to non-boolean.
+        // When one of the operand is not of boolean type, folding the
+        // other boolean constant will change return type of
+        // expression to non-boolean.
         //
         // Make sure c1 column to be used in tests is not boolean type
-        assert_eq!(col("c1").get_type(&schema)?, DataType::Utf8);
+        assert_eq!(col("c1").get_type(&schema).unwrap(), DataType::Utf8);
```

Comment:
Shall we add a test for nonboolean type for logical plan? Something like
```rust
  #[test]
    fn test_simplity_skip_nonboolean_type() {
        let table_scan = test_table_scan();
        let plan = LogicalPlanBuilder::from(table_scan)
            .filter(col("d").eq(lit(false)).not())
            .unwrap()
            .project(vec![col("a")])
            .unwrap()
            .build()
            .unwrap();

        let expected = "\
        Projection: #test.a\
        \n  Filter: NOT #test.d = Boolean(false)\
        \n    TableScan: test projection=None";

        assert_optimized_plan_eq(&plan, expected);
    }
```

Reply:
So I think that plan would error at runtime (because `"d"` is defined to be `DataTypre::UInt32` and if you try to do `UInt32 != Bool` you get an error: 

```
❯ select 1 != true;
Plan("'Int64 != Boolean' can't be evaluated because there isn't a common type to coerce the types to")
```


Though it does work if we have an explicit `CAST`:

```
❯ select 1 != cast(true as int);
+------------------------------------------+
| Int64(1) != CAST(Boolean(true) AS Int32) |
+------------------------------------------+
| false                                    |
+------------------------------------------+
1 row in set. Query took 0.003 seconds.
```

I guess I was thinking the plan tests are for ensuring that `simplify` is being called correctly on the expressions in the plan, rather than testing the various `Expr` corner cases of the simplification logic itself

Reply:
> I guess I was thinking the plan tests are for ensuring that simplify is being called correctly on the expressions in the plan, rather than testing the various Expr corner cases of the simplification logic itself

Thanks for the explanation. Makes sense to me.